packages/wallet/src/services/AddressDiscovery/MultiAddressDiscovery.ts

```diff
@@ -0,0 +1,80 @@
+import { AddressDiscovery } from '../types';
+import { AddressType, AsyncKeyAgent, GroupedAddress } from '@cardano-sdk/key-management';
+import { ChainHistoryProvider } from '@cardano-sdk/core';
+
+/**
+ * By default, we support up to five stake keys in the multi delegation schema.
+ */
+const DEFAULT_STAKE_KEY_INDEX_LIMIT = 5;
+
+/**
+ * Discover the addresses in the following manner:
+ *
+ * - Derive base addresses with payment credential at index 0 and increasing stake credential until it reaches the given limit.
+ * - Derives base addresses with increasing payment credential and stake credential at index 0.
+ * - if no transactions are found, increase the gap count.
+ * - if there are some transactions, increase the payment credential index and set the gap count to 0.
+ * - if the gap count reaches the given lookAheadCount stop the discovery process.
+ *
+ * Please note that the algorithm works with the transaction history, not balances, so you can have an address with 0 total coins
+ * and the algorithm will still continue with discovery if the address was previously used.
+ *
+ * If the wallet hits gap limit of unused addresses in a row, it expects there are
+ * no used addresses beyond this point and stops searching the address chain.
+ */
+export class MultiAddressDiscovery implements AddressDiscovery {
```

Comment:
Looks to me this is a sequential HD?
```suggestion
export class HDSequentialDiscovery implements AddressDiscovery {
```
https://input-output-hk.github.io/cardano-wallet/concepts/address-derivation